### PR TITLE
For #2346. Enable kotlin warningsAsErrors for `browser-domains` module.

### PR DESCRIPTION
**Pull request by dector**
### Issue #2346

  - Refactored (a bit) `ShippedDomainsProvider` and `CustomDomainsProvider`. Using
composition instead of inheritance. Was needed to cover with unit tests functionality in this classes. `Domains` loading already has tests, so `ShippedDomainsProvider` and `CustomDomainsProvider` can be represented as a building function if needed.
  - Add tests for `BaseDomainAutocompleteProvider`.

### Complexity

Easy+ (★☆☆)

### Pull Request checklist
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] ~~**Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features~~

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -12,6 +12,10 @@ permalink: /changelog/
 * [Gecko](https://github.com/mozilla-mobile/android-components/blob/master/buildSrc/src/main/java/Gecko.kt)
 * [Configuration](https://github.com/mozilla-mobile/android-components/blob/master/buildSrc/src/main/java/Config.kt)
 
+* **browser-domains**
+  * New domain autocomplete providers `ShippedDomainsProvider` and `CustomDomainsProvider` that
+    should be used instead of deprecated `DomainAutoCompleteProvider`.
```
Comment (pocmo) [inline docs/changelog.md]: nit: This needs to be moved to the latest version now. :)

File: components/browser/domains/src/main/java/mozilla/components/browser/domains/autocomplete/Providers.kt [hunk 2]
```diff
@@ -23,40 +23,40 @@ enum class DomainList(val listName: String) {
...
-    DomainAutocompleteProvider {
-    internal val scope = CoroutineScope(Dispatchers.IO)
-    // We compute 'domains' on the worker thread; make sure it's immediately visible on the UI thread.
-    @Volatile
+open class BaseDomainAutocompleteProvider(
+    private val list: DomainList,
+    private val domainsLoader: DomainsLoader
+) : DomainAutocompleteProvider, CoroutineScope by CoroutineScope(Dispatchers.IO) {
+
     var domains: List<Domain> = emptyList()
```
Comment (pocmo) [inline components/browser/domains/src/main/java/mozilla/components/browser/domains/autocomplete/Providers.kt]: Did you intentionally remove the `@Volatile` annotation here? With reading/writing the list from different threads it looks like we still need it?

File: docs/changelog.md (hunk 1 above)
Comment (dector) [reply docs/changelog.md]: It is in `0.54.0-SNAPSHOT`. :)

File: components/browser/domains/src/main/java/mozilla/components/browser/domains/autocomplete/Providers.kt (hunk 2 above)
Comment (dector) [reply components/browser/domains/src/main/java/mozilla/components/browser/domains/autocomplete/Providers.kt]: Yes, I've removed it intentionally but I was wrong doing this. :)
Will return it back.

**Comment by dector** [comment]
Moved changelog entry to 0.55.0

**Pull request merged**